### PR TITLE
feat: add SetScalingFactorController gov prop

### PR DESCRIPTION
**Pull request by sampocs**
Closes: #XXX

## What is the purpose of the change
Adds a gov prop to update the scaling factor controller address of a stableswap pool. This is particularly useful when migrating from a multisig controller to a CW contract controller.

### Setup
* Start localosmosis
* Modify the stableswapPool.json to include the scaling-factor-controller address
```json
{
        "initial-deposit": "1000000stake,1000uosmo",
        "swap-fee": "0.005",
        "exit-fee": "0",
        "future-governor": "",
        "scaling-factors": "1000,1",
        "scaling-factor-controller": "osmo19e2mf7cywkv7zaug6nk5f87d07fxrdgrladvymh2gwv5crvm3vnsuewhh7"
}
```
* Add controller key
```
echo "retire sport observe pyramid hole venue simple panel seven trophy group kiss shoe caution length farm sail main hunt client level hazard alien mirror" | osmosisd keys add controller --recover --keyring-backend test
osmosisd tx bank send osmo12smx2wdlyttvyzvzg54y2vnqwq2qjateuf7thj osmo13zfmpc7lr97ud4sg335wxt3e4x9x8zr7crvqyu 1000000uosmo --from val -y --keyring-backend test --chain-id localosmosis --fees 1000uosmo
```
* Create a stableswap pool
```bash
osmosisd tx gamm create-pool --pool-type stableswap --pool-file stableswapPool.json \
     --from val -y --keyring-backend test --chain-id localosmosis --fees 1000uosmo
```
* Show current address
```bash
>>> osmosisd q gamm pools | grep controller
  scaling_factor_controller: osmo19e2mf7cywkv7zaug6nk5f87d07fxrdgrladvymh2gwv5crvm3vnsuewhh7
```
### Update controller address with proposal.json
* Create proposal.json
```json
{
	"title": "Set Scaling Factor Controller Proposal",
	"description": "Change scaling factor controller address",
	"pool_id": 1,
	"controller_address": "osmo1jg7w3rt85tyj2hxky206chj23ts3nxm43gll3q"
}
```
* Submit proposal
```bash
osmosisd tx gov submit-proposal set-scaling-factor-controller-proposal \
    --proposal proposal.json  --deposit "10000000uosmo" \
    --from val -y --keyring-backend test --chain-id localosmosis --fees 1000uosmo
```
* Vote on proposal
```bash
osmosisd tx gov vote 1 yes --from val -y --keyring-backend test --chain-id localosmosis --fees 1000uosmo
```
* Wait for voting period to finish and confirm address updated
```bash
>>> osmosisd q gamm pools | grep controller
  scaling_factor_controller: osmo1jg7w3rt85tyj2hxky206chj23ts3nxm43gll3q
```

### Update controller address with flags
* Submit proposal with flags
```
osmosisd tx gov submit-proposal set-scaling-factor-controller-proposal \
      --title "Set Scaling Factor Controller Proposal" \
      --description "Change scaling factor controller address" \
      --deposit "10000000uosmo" \
      --pool-id 1 \
      --scaling-factor-controller-address "osmo13zfmpc7lr97ud4sg335wxt3e4x9x8zr7crvqyu" \
      --from val -y --keyring-backend test --chain-id localosmosis --fees 1000uosmo
```
* Vote on proposal
```bash
osmosisd tx gov vote 2 yes --from val -y --keyring-backend test --chain-id localosmosis --fees 1000uosmo
```
* Wait 1 minute for voting period to finish and confirm address updated
```bash
>>> osmosisd q gamm pools | grep controller
  scaling_factor_controller: osmo13zfmpc7lr97ud4sg335wxt3e4x9x8zr7crvqyu
```

### Adjust Scaling Factor
* Attempt to adjust the scaling factor with a non-controller address
```bash
osmosisd tx gamm adjust-scaling-factors --pool-id 1 --scaling-factors="2,1" --from val -y --keyring-backend test --chain-id localosmosis --fees 1000uosmo
```
* Query the tx and confirm that it failed with the error `not scaling factor governor`
* Attempt to adjust the scaling factor with the controller address
```bash
osmosisd tx gamm adjust-scaling-factors --pool-id 1 --scaling-factors="2,1" --from controller -y --keyring-backend test --chain-id localosmosis --fees 1000uosmo
```
* Confirm the scaling factor was updated
```bash
osmosisd q gamm pools | grep scaling_factors -A 2
```

### Error Cases
* Attempt to submit a proposal for a non-existent pool - query the tx and see that it failed because the pool ID does not exist
```bash
osmosisd tx gov submit-proposal set-scaling-factor-controller-proposal \
      --title "Set Scaling Factor Controller Proposal" \
      --description "Change scaling factor controller address" \
      --deposit "10000000uosmo" \
      --pool-id 100 \
      --scaling-factor-controller-address "osmo1w9e3xrjja7th9w9jje2vakghm8ppjnkrqrcm2t" \
      --from val -y --keyring-backend test --chain-id localosmosis --fees 1000uosmo
```
* Attempt to submit a proposal with an invalid address, it should fail
```bash
osmosisd tx gov submit-proposal set-scaling-factor-controller-proposal \
      --title "Set Scaling Factor Controller Proposal" \
      --description "Change scaling factor controller address" \
      --deposit "10000000uosmo" \
      --pool-id 1 \
      --scaling-factor-controller-address "osmoXXX" \
      --from val -y --keyring-backend test --chain-id localosmosis --fees 1000uosmo
```

## Documentation and Release Note
  - [X] Does this pull request introduce a new feature or user-facing behavior changes?
  - [X] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [X] Code comments?
  - [ ] N/A

### CODE REVIEW

File: x/gamm/handler.go [hunk 1]
```diff
@@ -11,14 +11,16 @@ import (
 	"github.com/osmosis-labs/osmosis/v17/x/gamm/types"
 )
 
-// NewMigrationRecordHandler is a handler for governance proposals on new migration records.
-func NewMigrationRecordHandler(k keeper.Keeper) govtypes.Handler {
+// NewGammProposalHandler is a handler for governance proposals on new migration records.
```
Comment (czarcas7ic) [inline x/gamm/handler.go:14]:
```suggestion
// NewGammProposalHandler is a handler for governance proposals for the GAMM module.
```

File: x/gamm/keeper/pool_test.go [hunk 2]
```diff
@@ -518,6 +518,83 @@ func (s *KeeperTestSuite) TestSetStableSwapScalingFactors() {
...
+	initialControllerAddr := s.TestAccs[0].String()
+	updatedControllerAddr := s.TestAccs[1].String()
+
+	testcases := []struct {
+		name             string
+		poolId           uint64
+		expError         error
+		isStableSwapPool bool
+	}{
+		{
```
Comment (czarcas7ic) [inline x/gamm/keeper/pool_test.go:531]: Lets add a test case where we try to use an invalid address as the controller addr

**Review by czarcas7ic — APPROVED** [review]
This LGTM! Lets get this tested on a local osmosis instance to be sure and then we can merge, thanks!

**Comment by czarcas7ic** [comment]
Oh and also could you please add a changelog entry as well?

**Comment by sampocs** [comment]
@czarcas7ic 
> Oh and also could you please add a changelog entry as well?

Added! (sorry forgot to push it up yesterday)

File: x/gamm/keeper/pool_test.go (hunk 2 above)
Comment (sampocs) [reply x/gamm/keeper/pool_test.go:531]: the address validation happens upstream in ValidateBasic. Want me to duplicate the check here as well?

**Comment by sampocs** [comment]
Successfully tested this locally with localosmosis and added the testing flow to the PR description

@czarcas7ic

File: x/gamm/types/codec.go [hunk 3]
```diff
@@ -23,6 +23,7 @@ func RegisterLegacyAminoCodec(cdc *codec.LegacyAmino) {
 	cdc.RegisterConcrete(&MsgExitSwapShareAmountIn{}, "osmosis/gamm/exit-swap-share-amount-in", nil)
 	cdc.RegisterConcrete(&UpdateMigrationRecordsProposal{}, "osmosis/gamm/update-migration-records-proposal", nil)
 	cdc.RegisterConcrete(&ReplaceMigrationRecordsProposal{}, "osmosis/gamm/replace-migration-records-proposal", nil)
+	cdc.RegisterConcrete(&SetScalingFactorControllerProposal{}, "osmosis/gamm/set-scaling-factor-controller-address", nil)
```
Comment (ValarDragon) [inline x/gamm/types/codec.go:26]: I think this is too long to work with ledgers, cc @mattverse isn't the limit 40 characters? Or does that not apply to gov props?

**Review by ValarDragon — APPROVED** [review]
utACK

Nice job! I think we need a different amino string due to the 40 char limit from the SDK. (It really should be in the API used for registration :/ , looks like some of our existing messags there have problems :/ )

File: x/gamm/types/codec.go (hunk 3 above)
Comment (sampocs) [reply x/gamm/types/codec.go]: Good catch! I think it's actually 39 characters. I'm not sure if it applies to gov props

Either way, shortened it to be safe [[here](https://github.com/osmosis-labs/osmosis/pull/5937/commits/35f088d8cad643d9c3722f123aafeacbaeb14c8e)]

**Pull request merged**